### PR TITLE
ipatests: Give the subCA more time to be loaded by the CA

### PR DESCRIPTION
**Pull request by rcritten**
ipatests: Give the subCA more time to be loaded by the CA

The subCA keys are loaded out-of-band after creation into the
CA so they may have been replicated but not loaded. Give more
time for them to appear in the remote CA.

Use a loop for the checking instead of a raw sleep because most
of the time this is very fast (< 15 seconds) but sometimes it
requires just a bit more. Allow up to 60 seconds.

Fixes: https://pagure.io/freeipa/issue/9096

Signed-off-by: Rob Crittenden <rcritten@redhat.com>

### CODE REVIEW

**Review by mrizwan93 — APPROVED** [review]
LGTM.

**Comment by flo-renaud** [comment]
Hi @rcritten,

the patch seems to fix the call to `self.check_subca` but unfortunately the test fails later (see my run using the nightly copr repo for pki: PR #[1550](https://github.com/freeipa-pr-ci2/freeipa/pull/1550)).

I think `certutil -K` is the culprit, since it sometimes returns the key nickname with `NSS Certificate DB:subsystemCert cert-pki-ca` or `subsystemCert cert-pki-ca` and the test is comparing the list of key names seen on master and replica:
```
INFO     ipatests.pytest_ipa.integration.host.Host.master.IPAOpenSSHTransport:transport.py:391 RUN ['/usr/bin/certutil', '-d', '/etc/pki/pki-tomcat/alias', '-K', '-f', '/etc/pki/pki-tomcat/alias/pwdfile.txt']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd76:transport.py:513 RUN ['/usr/bin/certutil', '-d', '/etc/pki/pki-tomcat/alias', '-K', '-f', '/etc/pki/pki-tomcat/alias/pwdfile.txt']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd76:transport.py:557 certutil: Checking token "NSS Certificate DB" in slot "NSS User Private Key and Certificate Services"
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd76:transport.py:557 < 0> rsa      091e15d85cd94e422e5e8fba3be0bba44ca3ddac   NSS Certificate DB:Server-Cert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd76:transport.py:557 < 1> rsa      d0149a5a8fe69b1fdf68caa3eeee152188443837   NSS Certificate DB:caSigningCert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd76:transport.py:557 < 2> rsa      8ea6c5d62884ce845d86bc44da9702df513455d4   NSS Certificate DB:ocspSigningCert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd76:transport.py:557 < 3> rsa      4b459fd24909269fd507dcb602c9322417c5b6ef   NSS Certificate DB:subsystemCert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd76:transport.py:557 < 4> rsa      a3e51f00aa55eb466188682238fa6635d41338ce   NSS Certificate DB:auditSigningCert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd76:transport.py:557 < 5> rsa      eb06ea91cb6480898c8931e960a4125aac6d6fd7   caSigningCert cert-pki-ca ef4bb985-5ab7-41e8-b675-24cd25acce4b
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd76:transport.py:217 Exit code: 0
INFO     ipatests.pytest_ipa.integration.host.Host.replica0.IPAOpenSSHTransport:transport.py:391 RUN ['/usr/bin/certutil', '-d', '/etc/pki/pki-tomcat/alias', '-L', '-f', '/etc/pki/pki-tomcat/alias/pwdfile.txt']
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:513 RUN ['/usr/bin/certutil', '-d', '/etc/pki/pki-tomcat/alias', '-L', '-f', '/etc/pki/pki-tomcat/alias/pwdfile.txt']
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557 
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557 Certificate Nickname                                         Trust Attributes
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557                                                              SSL,S/MIME,JAR/XPI
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557 
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557 caSigningCert cert-pki-ca                                    CTu,Cu,Cu
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557 ocspSigningCert cert-pki-ca                                  u,u,u
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557 auditSigningCert cert-pki-ca                                 u,u,Pu
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557 subsystemCert cert-pki-ca                                    u,u,u
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557 Server-Cert cert-pki-ca                                      u,u,u
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:557 caSigningCert cert-pki-ca ef4bb985-5ab7-41e8-b675-24cd25acce4b u,u,u
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd93:transport.py:217 Exit code: 0
INFO     ipatests.pytest_ipa.integration.host.Host.replica0.IPAOpenSSHTransport:transport.py:391 RUN ['/usr/bin/certutil', '-d', '/etc/pki/pki-tomcat/alias', '-K', '-f', '/etc/pki/pki-tomcat/alias/pwdfile.txt']
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd94:transport.py:513 RUN ['/usr/bin/certutil', '-d', '/etc/pki/pki-tomcat/alias', '-K', '-f', '/etc/pki/pki-tomcat/alias/pwdfile.txt']
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd94:transport.py:557 certutil: Checking token "NSS Certificate DB" in slot "NSS User Private Key and Certificate Services"
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd94:transport.py:557 < 0> rsa      d0149a5a8fe69b1fdf68caa3eeee152188443837   caSigningCert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd94:transport.py:557 < 1> rsa      8ea6c5d62884ce845d86bc44da9702df513455d4   ocspSigningCert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd94:transport.py:557 < 2> rsa      a3e51f00aa55eb466188682238fa6635d41338ce   auditSigningCert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd94:transport.py:557 < 3> rsa      4b459fd24909269fd507dcb602c9322417c5b6ef   subsystemCert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd94:transport.py:557 < 4> rsa      fe6333dc7cf0c4003e90676eea5706d1cebe3712   NSS Certificate DB:Server-Cert cert-pki-ca
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd94:transport.py:557 < 5> rsa      eb06ea91cb6480898c8931e960a4125aac6d6fd7   caSigningCert cert-pki-ca ef4bb985-5ab7-41e8-b675-24cd25acce4b
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd94:transport.py:217 Exit code: 0
```

**Comment by rcritten** [comment]
Latest patch, currently separate, to strip out the token name. We don't care what it is and it simplifies the FIPS checking as well.

**Comment by rcritten** [comment]
Ok so ignoring the tokens worked once. Whether it stands up I don't know.

**Comment by flo-renaud** [comment]
Hi @rcritten 
thanks for the updated patch, works for me. Tested using fedora 35 ([Report](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/8f2bd94a-9dea-11ec-90f5-fa163e2f4bc9/report.html)) and fedora 35 + pki nightly copr ([Report](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/c3ca556e-9dea-11ec-9f75-fa163ec53c61/report.html)), both work well.

You can remove the temp commit and either squash the 2 remaining commits into a single one, or fix the 2nd commit message (remove WIP).

**Comment by rcritten** [comment]
Temp commit dropped and the two changes squashed.

**Comment by flo-renaud** [comment]
@rcritten 
thanks for the update. ACK.

**Comment by stanislavlevin** [comment]
```diff
-        # give replication some time
-        time.sleep(15)
+        # give replication some time, up to 60 seconds
+        for _i in range(0,6):
+            time.sleep(10)
+            m = self.check_subca(master, self.SUBCA_MASTER, replica_nick)
+            r = self.check_subca(replica, self.SUBCA_MASTER, replica_nick)
+
+            if m and r:
+                break
+        else:
+            assert m, "master doesn't have the subCA"
+            assert r, "replica doesn't have the subCA"
 
         # replica.run_command(['ipa-certupdate'])
-        self.check_subca(replica, self.SUBCA_REPLICA, replica_nick)
-        self.check_subca(master, self.SUBCA_REPLICA, replica_nick)
```
typo `self.SUBCA_MASTER` => `self.SUBCA_REPLICA` in `test_subca_replica`.

**Comment by rcritten** [comment]
> typo `self.SUBCA_MASTER` => `self.SUBCA_REPLICA` in `test_subca_replica`.

Yeoch. Good catch.

Re-running temp commit to ensure it still works. The previous code just re-checked that the master-created key replicated.

**Comment by stanislavlevin** [comment]
I hit this issue last week on migrating dogtag-pki and its subcomponents from 11.0.3 to 11.1.0.

I run my IPA CI on top of this PR and can confirm that the issue looks fixed.
@rcritten, thank you.

File: ipatests/prci_definitions/temp_commit.yaml [hunk 1]
```diff
@@ -68,7 +68,7 @@ jobs:
       class: RunPytest
       args:
         build_url: '{fedora-latest/build_url}'
-        test_suite: test_integration/test_REPLACEME.py
+        test_suite: test_integration/test_replica_promotion.py::TestSubCAkeyReplication
         template: *ci-master-latest
-        timeout: 3600
-        topology: *master_1repl_1client
+        timeout: 720
```
Comment (flo-renaud) [inline ipatests/prci_definitions/temp_commit.yaml:73]: Typo, 720 instead of 7200

**Comment by rcritten** [comment]
Temp commit removed.

**Comment by flo-renaud** [comment]
master:

* 9ac88216a00130632a6bd7c7732d8bc51eab4ba1 ipatests: Give the subCA more time to be loaded by the CA

**Pull request closed without merging**